### DISCUSSION TITLE
Reactice watch is not necessarily called when the underlying object changes

### DISCUSSION THREAD

**Discussion by samwareing** [Textualize/textual#2193] [Q&A]
This issue occurs because of the assumption that the object a `reactive` watches will only change value when the `__set__` operator is called on the `reactive`.

If I have an object that has a deeper structue, or can be changed by calling a function in some way, I would expect `reactive` to know that it something has changed and call the watch function.

In the example below, I have created a `reactive` that watches a `dataclass` object that is able to change itself via a the `initialise_my_types` method.

I then call that method via a `textual` action, and I can see the underlying value change in my debugger, but the watch function is not called and therefore the displayed value is out of date.

```python
from dataclasses import dataclass, field

from textual.app import App, ComposeResult
from textual.containers import Container
from textual.reactive import reactive
from textual.widgets import Footer, Header, Static


@dataclass
class MyType:
    id: int = 0


@dataclass
class DataclassExample:
    my_types: list[MyType] = field(default_factory=list)

    def initialise_my_types(self) -> None:
        self.my_types.append(MyType())
        pass


class Display(Static):
    dce = reactive(DataclassExample())

    def watch_dce(self, dce: DataclassExample) -> None:
        self.update(f"{self.dce!r}")

    def initialise_my_types(self) -> None:
        self.dce.initialise_my_types()
        # a workaround is to call self.update() here


class ExampleApp(App):
    BINDINGS = [("i", "initialise_my_type", "Initialise example")]

    def compose(self) -> ComposeResult:
        yield Container(Display())
        yield Footer()

    def action_initialise_my_type(self) -> None:
        self.query_one(Display).initialise_my_types()


def main():
    app = ExampleApp()
    app.run()


if __name__ == "__main__":
    main()
```

The following screenshots show the value changing. There is a breakpoint in the watch function that is never hit.

<img width="1512" alt="SCR-20230401-ktyi" src="https://user-images.githubusercontent.com/98537672/229285067-4ef78601-9965-4fe5-9058-ed38cca1ac0a.png">

<img width="1512" alt="SCR-20230401-kuar" src="https://user-images.githubusercontent.com/98537672/229285070-4192559b-5702-4356-a9c8-955d17fac9a9.png">

# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.17.1 |
| Rich    | 13.3.3 |

## Python

| Name           | Value                                                       |
|----------------|-------------------------------------------------------------|
| Version        | 3.10.9                                                      |
| Implementation | CPython                                                     |
| Compiler       | Clang 14.0.0 (clang-1400.0.29.202)                          |
| Executable     | /Users/samwareing/code/textual/practice/venv/bin/python3.10 |

## Operating System

| Name    | Value                                                                                                |
|---------|------------------------------------------------------------------------------------------------------|
| System  | Darwin                                                                                               |
| Release | 22.3.0                                                                                               |
| Version | Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000 |

## Terminal

| Name                 | Value           |
|----------------------|-----------------|
| Terminal Application | vscode (1.77.0) |
| TERM                 | xterm-256color  |
| COLORTERM            | truecolor       |
| FORCE_COLOR          | *Not set*       |
| NO_COLOR             | *Not set*       |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=125, height=12 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 125                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 12                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

```

**Comment by willmcgugan** [comment]
That's the expected behaviour at the moment. Reactives will only detect changes when you set a new value, not when you mutate an existing value.

You can do `self.my_types == self.my_types` to prompt the watchers.

**Reply by samwareing** [reply]
Hi Will,

I actually tried this last week, it still does not work for this example.

I tried it again though, here's my change:

```python
class Display(Static):
    dce = reactive(DataclassExample())

    def watch_dce(self, dce: DataclassExample) -> None:
        self.update(f"{self.dce!r}")

    def initialise_my_types(self) -> None:
        self.dce.initialise_my_types()
        self.dce = self.dce
```

I think the below screenshot shows why the watcher is not prompted:

<img width="1512" alt="SCR-20230401-lnlt" src="https://user-images.githubusercontent.com/98537672/229286981-37af3f44-027e-4bf0-9843-e9c61a76e5ad.png">

`current_value` already has the change.

A workaround that does work is to set `always_update` true.

edit: scratch that, I misread what you typed, will look at settings `self.my_types`.

**Reply by samwareing** [reply]
Hmm, still no joy with this:

```python

@dataclass
class DataclassExample:
    my_types: list[MyType] = field(default_factory=list)

    def initialise_my_types(self) -> None:
        self.my_types.append(MyType())
        self.my_types = self.my_types


class Display(Static):
    dce = reactive(DataclassExample())

    def watch_dce(self, dce: DataclassExample) -> None:
        self.update(f"{self.dce!r}")

    def initialise_my_types(self) -> None:
        self.dce.initialise_my_types()
        self.dce = self.dce
```
